dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java

```diff
@@ -0,0 +1,343 @@
+        String publicClassName = getPublicClassName(sourceCode);
+        String fileName =  buildJavaSourceCodeFileFullName(publicClassName);
+        createJavaSourceFileIfNotExists(sourceCode, fileName);
+        String compileCommand = buildJavaCompileCommand(fileName, sourceCode);
+        Preconditions.checkNotNull(compileCommand, "command not be null.");
+        TaskResponse compileResponse = shellCommandExecutor.run(compileCommand);
+        // must drop the command file ,if do not ,the next command will not run. because be limited the ShellCommandExecutor's create file rules
+        dropShellCommandFile();
+        shellCommandExecutor = new ShellCommandExecutor(this::logHandle,
+                taskRequest,
+                logger);
+        logger.info("java task code compile result : " + compileResponse);
+        return publicClassName;
+    }
+
+    private void dropShellCommandFile() throws IOException {
+        String commandFilePath = String.format("%s/%s.%s"
+                , taskRequest.getExecutePath()
+                , taskRequest.getTaskAppId()
+                , SystemUtils.IS_OS_WINDOWS ? "bat" : "command");
+        Path path = Paths.get(commandFilePath);
+        if (Files.exists(path)) {
+            Files.delete(path);
+        }
+    }
+
+    protected String buildJavaCompileCommand(String fileName, String sourceCode) throws IOException {
+
+        StringBuilder compilerCommand = new StringBuilder()
+                .append(getJavaHomeBinAbsolutePath())
+                .append("javac").append(" ")
+                .append(buildResourcePath()).append(" ")
+                .append(fileName);
+        return compilerCommand.toString();
+    }
+
+    protected String buildJavaSourceContent() {
+        String rawJavaScript = javaParameters.getRawScript().replaceAll("\\r\\n", "\n");
+        // replace placeholder
+
+        Map<String, Property> paramsMap = taskRequest.getPrepareParamsMap();
+        if (MapUtils.isEmpty(paramsMap)) {
+            paramsMap = new HashMap<>();
+        }
+        if (MapUtils.isNotEmpty(taskRequest.getParamsMap())) {
+            paramsMap.putAll(taskRequest.getParamsMap());
+        }
+        rawJavaScript = ParameterUtils.convertParameterPlaceholders(rawJavaScript, ParamUtils.convert(paramsMap));
+        logger.info("raw java script : {}", javaParameters.getRawScript());
+        return rawJavaScript;
+    }
+
+    private String getJavaHomeBinAbsolutePath() {
+        String javaHomeAbsolutePath = System.getenv(JavaConstants.JAVA_HOME);
+        Preconditions.checkNotNull(javaHomeAbsolutePath, "not find the java home in the version. ");
+        return javaHomeAbsolutePath + System.getProperty("file.separator") + "bin" + System.getProperty("file.separator");
+    }
+
+    public String getPublicClassName(String sourceCode) {
+        String pattern = "(.*\\s+public\\s+class\\s+)([a-zA-Z_]+[//w_]*)([.\\s\\S]*)";
+        Pattern compile = Pattern.compile(pattern);
```

Comment:
I think should cache regex pattern to improve performance, its performance matters.

Reply:
good idea, have done.